### PR TITLE
Read/Write layer

### PR DESCRIPTION
**Pull request by achimweigel**
**How to categorize this PR?**

/area operations
/kind enhancement
/priority 3

**What this PR does / why we need it**:

This PR adds a read and write layer for all read and write operations on installations, executions and deploy items. This layer has two goals:

- allows to investigate the controller logic by string at these methods and using the call hierarchy navigation
- is the prerequisite to add information about who modified what to the objects 

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```other operator
NONE
```

### CODE REVIEW

File: pkg/landscaper/execution/reconcile.go [hunk 1]
```diff
@@ -200,7 +201,7 @@ func (o *Operation) deployOrTrigger(ctx context.Context, item executionItem) err
 	o.exec.Status.DeployItemReferences = lsv1alpha1helper.SetVersionedNamedObjectReference(o.exec.Status.DeployItemReferences, ref)
 	o.exec.Status.ExecutionGenerations = setExecutionGeneration(o.exec.Status.ExecutionGenerations, item.Info.Name, o.exec.Generation)
 	o.exec.Status.Phase = lsv1alpha1.ExecutionPhaseProgressing
-	if err := o.Client().Status().Patch(ctx, o.exec, client.MergeFrom(old)); err != nil {
+	if err := read_write_layer.PatchExecution(ctx, o.Client(), o.exec, old); err != nil {
```
Comment (robertgraeff) [inline pkg/landscaper/execution/reconcile.go:204]: The **status** of the execution must be patched.

Comment (robertgraeff) [inline pkg/landscaper/execution/reconcile.go:204]: The **status** of the execution must be patched.

**Review by robertgraeff — CHANGES_REQUESTED** [review]

File: pkg/landscaper/execution/delete.go [hunk 2]
```diff
@@ -61,7 +63,7 @@ func (o *Operation) Delete(ctx context.Context) error {
 	}
 
 	controllerutil.RemoveFinalizer(o.exec, lsv1alpha1.LandscaperFinalizer)
-	return lserrors.NewErrorOrNil(o.Client().Update(ctx, o.exec), op, "RemoveFinalizer")
+	return lserrors.NewErrorOrNil(read_write_layer.UpdateExecutionStatus(ctx, o.Client(), o.exec), op, "RemoveFinalizer")
```
[the remark this answers is not in the document]
Comment (achimweigel) [reply pkg/landscaper/execution/delete.go:66]: good finding

File: pkg/landscaper/execution/reconcile.go (hunk 1 above)
Comment (achimweigel) [reply pkg/landscaper/execution/reconcile.go:204]: yes that's another good finding

File: pkg/utils/read_write_layer/read.go [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+func ListExecutions(ctx context.Context, c client.Reader, executions *lsv1alpha1.ExecutionList, opts ...client.ListOption) error {
+	return list(ctx, c, executions, opts...)
+}
+
+// read methods for deploy items
+func GetDeployItem(ctx context.Context, c client.Reader, key client.ObjectKey, deployItem *lsv1alpha1.DeployItem) error {
+	return get(ctx, c, key, deployItem)
+}
+
+func ListDeployItem(ctx context.Context, c client.Reader, deployItems *lsv1alpha1.DeployItemList, opts ...client.ListOption) error {
```
[the remark this answers is not in the document]
Comment (achimweigel) [reply pkg/utils/read_write_layer/read.go:36]: I have renamed it accordingly

**Review by robertgraeff — APPROVED** [review]

**Pull request merged**